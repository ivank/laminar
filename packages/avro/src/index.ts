/**
 * @packageDocumentation
 * @module @laminar/avro
 */
export { AvroDecimal, bufferToDecimal, decimalToBuffer, to64Bit } from './avro-decimal';
export { AvroTimestampMillis } from './avro-timestamp-millis';
export { AvroEpochDays } from './avro-epoch-days';
