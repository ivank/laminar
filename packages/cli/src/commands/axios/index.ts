import { Command, createCommand, Option } from 'commander';
import fs from 'fs';
import { printDocument } from '@laminar/ts-compose';
import { convertOapi } from './convert';
import chalk from 'chalk';
import { dirname } from 'path';
import { Logger } from '../../types';
import { toContext, concatStreamToString, parseSchema } from '../../helpers';

export const processFile = async (file: string): Promise<string> => {
  const { context, value } = await toContext(file);
  return printDocument(convertOapi(context, value));
};

export const axiosCommand = (logger: Logger = console): Command =>
  createCommand('axios')
    .description('Convert openapi schemas to typescript types for axios')
    .option('-f, --file <file>', 'Filename to process, uses STDIN if not specified')
    .addOption(new Option('-t, --stdin-type <type>', 'Type of STDIN data').default('json').choices(['json', 'yaml']))
    .option('-o, --output <output>', 'File to output to, uses STDOUT if not specified')
    .action(async ({ file, output, stdinType }) => {
      const input = file ? file : parseSchema(stdinType, await concatStreamToString(process.stdin));
      const result = await processFile(input);

      if (output) {
        fs.mkdirSync(dirname(output), { recursive: true });
        logger.info(`OpanAPI Schema ${chalk.green(file ? file : 'STDIN')} -> ${chalk.yellow(output)} axios types`);
        fs.writeFileSync(output, result);
      } else {
        process.stdout.write(result);
      }
    });
