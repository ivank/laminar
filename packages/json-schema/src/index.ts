/**
 * @packageDocumentation
 * @module @laminar/json-schema
 */

export { ValidationError } from './ValidationError';
export { ResolveError } from './ResolveError';
export {
  InvalidCode,
  Invalid,
  Messages,
  Options,
  Validator,
  Validation,
  error,
  empty,
  onlyErrors,
  errors,
  evaluateItem,
  evaluateProperty,
  skipRest,
  hasErrors,
  hasEvaluatedItem,
  hasEvaluatedProperty,
  combine,
  combineErrors,
  childOptions,
  validateSchema,
} from './validation';
export { Schema, JsonSchema, PrimitiveType } from './schema';
export { toMessages, toMessage, messages } from './messages';
export {
  isCompiled,
  withinContext,
  toSchemaObject,
  compile,
  ResolvedSchema,
  resolve,
  resolveFile,
  RefMap,
  FileContext,
} from './resolve';
export {
  validate,
  validateCompiled,
  ensureValid,
  ensureValidCompiled,
  ValidateOptions,
  ValidateSchemaOptions,
  ValidateCompiledOptions,
  Result,
  ResultError,
  ResultSuccess,
  Drafts,
  FormatErrors,
} from './validate';
export { CoerceCompiledOptions, CoerceSchemaOptions, CoerceOptions, coerce, coerceCompiled } from './coerce';
export { coerceSchema, coercers, CoercerOptions, Coercer } from './coercion';
