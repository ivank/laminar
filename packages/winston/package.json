{
  "name": "@laminar/winston",
  "version": "0.15.0",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "description": "Use winston with @laminar/laminar",
  "author": "Ivan Kerin <ikerin@gmail.com>",
  "repository": "git@github.com:ivank/laminar.git",
  "homepage": "https://github.com/ivank/laminar/tree/main/packages/winston#readme",
  "license": "Apache-2.0",
  "devDependencies": {
    "@ikerin/build-readme": "^1.1.1",
    "@laminar/laminar": "workspace:*",
    "@types/jest": "^29.5.8",
    "@types/node": "^20.9.0",
    "@typescript-eslint/eslint-plugin": "^6.11.0",
    "@typescript-eslint/parser": "^6.11.0",
    "axios": "^1.6.2",
    "eslint": "^8.53.0",
    "eslint-config-prettier": "^9.0.0",
    "jest": "^29.7.0",
    "prettier": "^3.1.0",
    "ts-jest": "^29.1.1",
    "ts-retry-promise": "^0.7.1",
    "typescript": "^5.2.2",
    "winston": "^3.11.0"
  },
  "scripts": {
    "build": "tsc --declaration",
    "build:readme": "build-readme README.md https://github.com/ivank/laminar/tree/main/packages/winston",
    "test": "jest test --runInBand",
    "lint": "F='{src,test}/**/*.ts' && prettier --check $F && eslint $F"
  },
  "jest": {
    "preset": "../../jest.json"
  },
  "peerDependencies": {
    "@laminar/laminar": "*",
    "winston": "*"
  }
}
