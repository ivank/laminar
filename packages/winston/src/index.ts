/**
 * @packageDocumentation
 * @module @laminar/winston
 */
export { WinstonService } from './winston.service';
