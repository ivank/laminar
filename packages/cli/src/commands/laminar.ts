import commander from 'commander';
import { apiCommand } from './api';
import { axiosCommand } from './axios';
import { jsonSchemaCommand } from './json-shema';
import { convertCommand } from './avro';
import { Logger } from '../types';
import { fixturesCommand } from './fixtures';

export const laminarCommand = (output: Logger = console): commander.Command =>
  commander
    .createCommand('laminar')
    .version('0.15.0')
    .description(
      `Laminar CLI - a tool for converting OpenApi and Avro Schemas to appropriate typescript types

laminar api
  Convert openapi schema to types used by laminar api (openApiTyped middleware)

laminar axios
  Convert openapi schema to types used by axios

laminar json-schema
  Convert json schema to typescript types

laminar avro
  Convert avro to typescript types

laminar fixtures
  Load pg tables and generate @laminar/fixtures types from them

Example:
  laminar api --file openapi.yaml --output src/__generated__/openapi.yaml.ts
  curl http://example.com/openapi.yaml | laminar api --stdin-type yaml | prettier --stdin-filepath openapi.yamlts > src/__generated__/openapi.yaml.ts
  laminar api --file openapi.yaml --output src/__generated__/openapi.yaml.ts --watch
  laminar avro path/to/*.avsc --output-dir src/__generated__ --logical-type timestamp-millis=Date
  laminar fixtures postgres://example-admin:example-pass@localhost:5432/example --output src/__generated__/fixtures.ts
`,
    )
    .addCommand(apiCommand(output))
    .addCommand(axiosCommand(output))
    .addCommand(jsonSchemaCommand(output))
    .addCommand(convertCommand(output))
    .addCommand(fixturesCommand(output));
