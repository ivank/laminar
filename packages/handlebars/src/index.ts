/**
 * @packageDocumentation
 * @module @laminar/handlebars
 */

import { Middleware } from '@laminar/laminar';
import { readdirSync, readFileSync, existsSync } from 'fs';
import { compile, RuntimeOptions, TemplateDelegate } from 'handlebars';
import { join, relative } from 'path';
import { statSync } from 'fs';

export type Cache = Map<string, { template: TemplateDelegate; mtime: Date }>;

/**
 * Cache types.
 * - preload: load all the templates ones into memory
 * - expiry: load partials when needed and keep them in cache, but check the file's mtime and reload template if changed
 * - none: do not cache templates and load them on every request.
 */
export type CacheType = 'preload' | 'expiry' | 'none';

export interface TemplateConfig {
  /**
   * Helper functions passed to handlebar templates
   */
  helpers?: RuntimeOptions['helpers'];
  /**
   * Main directory where handlebar files should be located, defaults to current working directory.
   * Looks for "partials" and "views" directories inside of it.
   */
  dir?: string;
  /**
   * Directory where all the partials are located, inside the main 'dir' directory, defaults to 'partials'
   */
  partials?: string;
  /**
   * Name of the handlebars file extension. Defaults to 'handlebars', so that only 'my-file.handlebars' kind of files are loaded.
   */
  extension?: string;
  /**
   * Directory name where all the views are located, inside the main 'dir' directory, defaults to 'views'
   */
  views?: string;
  /**
   * Cache types.
   * - preload: load all the templates ones into memory
   * - expiry: load partials when needed and keep them in cache, but check the file's mtime and reload template if changed
   * - none: do not cache templates and load them on every request.
   *
   * Defaults to 'preload'
   */
  cacheType?: CacheType;
  /**
   * Global data to be passed to every template
   */
  globalData?: Record<string, unknown>;
}

export interface CompileTemplatesOptions {
  dir: string;
  extension: string;
}

export type HandlebarsRender = (view: string, data?: Record<string, unknown>) => string;

export interface HandlebarsContext {
  hbs: HandlebarsRender;
}

const hasExtension =
  (extension: string) =>
  (file: string): boolean =>
    file.endsWith(extension);

export const deepReaddirSync = ({ dir, extension }: CompileTemplatesOptions): string[] =>
  readdirSync(dir, { withFileTypes: true })
    .reduce<string[]>((all, item) => {
      return all.concat(
        item.isDirectory() ? deepReaddirSync({ dir: join(dir, item.name), extension }) : join(dir, item.name),
      );
    }, [])
    .filter(hasExtension(extension));

const toName = (dir: string, file: string): string => relative(dir, file.slice(0, file.lastIndexOf('.')));

export const compileTemplates = ({ dir, extension }: CompileTemplatesOptions): Array<[string, TemplateDelegate]> =>
  deepReaddirSync({ dir, extension }).reduce<Array<[string, TemplateDelegate]>>(
    (all, file) => [...all, [toName(dir, file), compile(readFileSync(file, 'utf8'))]],
    [],
  );

export const loadTemplate = (
  name: string,
  cache: Cache,
  cacheType: CacheType,
  { dir, extension }: CompileTemplatesOptions,
): TemplateDelegate => {
  const file = `${join(dir, name)}.${extension}`;

  const stat = cacheType === 'expiry' ? statSync(file) : { mtime: new Date(0) };
  const cached = cache.get(name);

  if (cacheType !== 'none' && cached && stat.mtime < cached.mtime) {
    return cached.template;
  } else {
    const template = compile(readFileSync(file, 'utf8'));
    cache.set(name, { template, mtime: cacheType === 'expiry' ? stat.mtime : new Date() });
    return template;
  }
};

const loadPartials = (options: CompileTemplatesOptions): { [key: string]: TemplateDelegate } =>
  existsSync(options.dir)
    ? compileTemplates(options).reduce((acc, [name, template]) => ({ ...acc, [name]: template }), {})
    : {};

const loadTemplates = (cache: Cache, options: CompileTemplatesOptions): Cache =>
  existsSync(options.dir)
    ? compileTemplates(options).reduce(
        (cache, [name, template]) => cache.set(name, { template, mtime: new Date() }),
        cache,
      )
    : cache;

export const handlebars = ({
  helpers = {},
  dir = process.cwd(),
  extension = 'handlebars',
  partials = 'partials',
  views = 'views',
  cacheType = 'preload',
  globalData = {},
}: TemplateConfig): HandlebarsRender => {
  const cache = cacheType === 'preload' ? loadTemplates(new Map(), { dir, extension }) : new Map();
  const partialTemplates = cacheType === 'preload' ? loadPartials({ dir: join(dir, partials), extension }) : undefined;

  return (name: string, data: Record<string, unknown> = {}) => {
    const template = loadTemplate(name, cache, cacheType, { dir: join(dir, views), extension });
    return template(
      { ...globalData, ...data },
      { helpers, partials: partialTemplates ?? loadPartials({ dir: join(dir, partials), extension }) },
    );
  };
};

export const handlebarsMiddleware = (config: TemplateConfig): Middleware<HandlebarsContext> => {
  const hbs = handlebars(config);
  return (next) => (req) => next({ ...req, hbs });
};
