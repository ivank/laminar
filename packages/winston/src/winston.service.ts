/* eslint-disable @typescript-eslint/no-explicit-any */
import { Service, LoggerLike, LoggerMetadata } from '@laminar/laminar';
import type { Logger } from 'winston';

/**
 * A {@link Service} wrapper for a winston logger instance to manage its lifecycle
 *
 * Forwards all calls to `debug`, `info`, `warn`, `error` to the winston instance.
 *
 * Calls logger.close() on stop
 */
export class WinstonService implements Service, LoggerLike {
  constructor(public source: Logger) {}

  debug(message: any, metadata?: LoggerMetadata): void {
    this.source.log({ level: 'debug', message, ...metadata });
  }

  info(message: any, metadata?: LoggerMetadata): void {
    this.source.log({ level: 'info', message, ...metadata });
  }

  warn(message: any, metadata?: LoggerMetadata): void {
    this.source.log({ level: 'warn', message, ...metadata });
  }

  error(message: any, metadata?: LoggerMetadata): void {
    this.source.log({ level: 'error', message, ...metadata });
  }

  async start(): Promise<this> {
    return this;
  }

  async stop(): Promise<this> {
    this.source.close();
    return this;
  }

  describe(): string {
    return 'ðŸ“œ Winston Logger';
  }
}
