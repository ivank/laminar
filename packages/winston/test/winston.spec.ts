import axios from 'axios';
import { HttpService, textOk, Middleware, run } from '@laminarjs/laminar';
import { SimpleQueue, Boss } from './simple-queue';
import { WinstonService } from '../src';
import { createLogger, transports } from 'winston';
import { join } from 'path';
import { readFileSync, existsSync, unlinkSync } from 'fs';

export interface BossContext<TData> {
  boss: Boss<TData>;
}
export const bossMiddleware =
  <TData>(boss: Boss<TData>): Middleware<BossContext<TData>> =>
  (next) =>
  (ctx) =>
    next({ ...ctx, boss });

describe('Services', () => {
  jest.setTimeout(10000);
  it('Should start and stop services', async () => {
    const logFile = join(__dirname, 'test.log');
    if (existsSync(logFile)) {
      unlinkSync(logFile);
    }
    const port = 8200;
    const boss = new Boss<number>();
    const winston = createLogger({ transports: new transports.File({ filename: logFile }) });
    const logger = new WinstonService(winston);

    const withBoss = bossMiddleware(boss);

    const queue = new SimpleQueue<number>(boss, [
      {
        queue: 'one',
        app: async ({ data }) => logger.info('Data', { data }),
      },
    ]);

    const http = new HttpService({
      port,
      listener: withBoss(async ({ boss, query: { data } }) => {
        logger.info('test');

        for (const item of data) {
          boss.add('one', Number(item));
        }
        return textOk('OK');
      }),
    });

    await run(
      {
        initOrder: [boss, [http, queue]],
        logger,
      },
      async () => {
        await axios.get(`http://localhost:${port}?data[]=1&data[]=2&data[]=3`);
        await axios.get(`http://localhost:${port}?data[]=4`);

        await new Promise((resolve) => setTimeout(resolve, 100));
      },
    );

    await new Promise((resolve) => setTimeout(resolve, 1000));

    const logContents = readFileSync(logFile, 'utf-8');

    expect(logContents).toMatchSnapshot();
  });
});
