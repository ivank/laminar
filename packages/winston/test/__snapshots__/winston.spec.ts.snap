// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Services Should start and stop services 1`] = `
"{"level":"info","message":"⏫ Starting Boss"}
{"level":"info","message":"✅ Started Boss"}
{"level":"info","message":"⏫ Starting ⛲ Laminar: http://0.0.0.0:8200"}
{"level":"info","message":"⏫ Starting Queue: one"}
{"level":"info","message":"✅ Started Queue: one"}
{"level":"info","message":"✅ Started ⛲ Laminar: http://0.0.0.0:8200"}
{"level":"info","message":"test"}
{"data":1,"level":"info","message":"Data"}
{"data":2,"level":"info","message":"Data"}
{"data":3,"level":"info","message":"Data"}
{"level":"info","message":"test"}
{"data":4,"level":"info","message":"Data"}
{"level":"info","message":"⏬ Stopping ⛲ Laminar: http://0.0.0.0:8200"}
{"level":"info","message":"⏬ Stopping Queue: one"}
{"level":"info","message":"❎ Stopped Queue: one"}
{"level":"info","message":"❎ Stopped ⛲ Laminar: http://0.0.0.0:8200"}
{"level":"info","message":"⏬ Stopping Boss"}
{"level":"info","message":"❎ Stopped Boss"}
{"level":"info","message":"❎ Stop 📜 Winston Logger"}
"
`;
